pkg/objstore/backoff.go

```diff
@@ -0,0 +1,248 @@
+package objstore
+
+import (
+	"context"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+type backoffBucket struct {
+	Bucket
```

Comment:
I would make it a named field. The problem is that whenever we'll add more functions to `Bucket` we may want this implementation to break so we can take an informed decision whether to add retries or not to the new function.